### PR TITLE
update tests for K8s 1.22+ and remove support for K8s < 1.19

### PR DESCRIPTION
**Pull request by jpfourny**
**What are you trying to accomplish with this PR?**
Have CI tests pass for K8s 1.22+ and truncate supported K8s version to minimum of 1.19.

The overall goal was to produce a single set of fixtures that are compatible with all supported version of K8s [1.19, 1.23]. It would introduce too much complexity into the test framework to maintain multiple fixture versions and apply them conditionally based on K8s server version in context. 

**How is this accomplished?**

- Update or remove test fixtures for resources that are no longer support in K8s 1.22
- `MutatingWebhookConfiguration` no longer supports the `Unknown` and `Some` side effects. This breaks tests against K8s 1.22 that exercise the batch dry-runner with dry-run unsafe hooks. For context, see https://github.com/Shopify/krane/pull/798. As long as we support K8s < 1.22, we should continue testing this logic. As a workaround, I stubbed the `has_side_effects?` to simulate pre-1.22 behaviour. 

**What could go wrong?**
Only tests should be affected. Although `krane` should still work with older K8s servers, we will no longer validate them in CI.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -11,24 +11,23 @@ jobs:
       fail-fast: false
       matrix:
         ruby:
-        - '3.0' # With k8s 1.23
-        - '3.0' # With k8s 1.22
-        - '3.0' # With k8s 1.21
+        # Use unique Ruby versions, or GitHub gets confused when building the rest of the matrix
```
Comment (jpfourny) [inline .github/workflows/ci.yml:14]: Ported from https://github.com/Shopify/krane/pull/860/files, except truncated k8s versions to >=1.19.

File: dev.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ up:
   - custom:
       name: Create KinD Cluster
       met?: bin/kind get clusters | grep -q krane
-      meet: bin/kind create cluster --name krane
+      meet: bin/kind create cluster --name krane --image "kindest/node:v1.22.0@sha256:b8bda84bb3a190e6e028b1760d277454a72267a5454b57db34437c34a588d047"
```
Comment (jpfourny) [inline dev.yml:16]: Dev env should test with k8s that Shopify is currently using in production.

File: dev.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ---
 name: krane
 up:
-  - ruby: 2.6.6 # Matches gemspec
+  - ruby: '3.0.3' # Matches gemspec
```
Comment (jpfourny) [inline dev.yml:4]: Any supported version will do for dev env, but Ruby 3 is a bit faster.

File: test/fixtures/crd/mail.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-apiVersion: apiextensions.k8s.io/v1beta1
+apiVersion: apiextensions.k8s.io/v1
```
Comment (jpfourny) [inline test/fixtures/crd/mail.yml:1]: `v1beta1` dropped in k8s `1.22`

File: test/fixtures/crd/mail.yml [hunk 5]
```diff
@@ -12,4 +12,16 @@ spec:
     plural: mail
     singular: mail
   scope: Namespaced
-  version: v1
+  versions:
```
Comment (jpfourny) [inline test/fixtures/crd/mail.yml:15]: `version` no longer supported; replaced by `versions`.

File: test/fixtures/crd/mail.yml [hunk 6]
```diff
@@ -12,4 +12,16 @@ spec:
...
+    - name: v1
+      served: true
+      storage: true
+      schema:
```
Comment (jpfourny) [inline test/fixtures/crd/mail.yml:19]: `schema` is now required.

File: test/fixtures/crd/redis.yml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-apiVersion: apiextensions.k8s.io/v1beta1
+apiVersion: apiextensions.k8s.io/v1
```
Comment (jpfourny) [inline test/fixtures/crd/redis.yml:1]: `v1beta1` dropped in k8s `1.22`

File: test/fixtures/crd/web.yml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-apiVersion: extensions/v1beta1
+apiVersion: networking.k8s.io/v1
```
Comment (jpfourny) [inline test/fixtures/crd/web.yml:1]: `extensions/v1beta1` dropped in k8s `1.22`; replaced with `networking.k8s.io/v1`

File: test/fixtures/crd/mail.yml [hunk 9]
```diff
@@ -12,4 +12,16 @@ spec:
     plural: mail
     singular: mail
   scope: Namespaced
-  version: v1
+  versions:
+    - name: v1
+      served: true
```
Comment (jpfourny) [inline test/fixtures/crd/mail.yml:17]: `served` and `storage` are mandatory

File: test/fixtures/crd/with_default_conditions.yml [hunk 10]
```diff
@@ -14,4 +14,31 @@ spec:
     plural: parameterizeds
     singular: parameterized
   scope: Namespaced
-  version: v1
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
```
Comment (jpfourny) [inline test/fixtures/crd/with_default_conditions.yml:22]: Schema inferred from tests

File: test/unit/cluster_resource_discovery_test.rb [hunk 11]
```diff
@@ -14,7 +14,7 @@ def test_fetch_resources_failure
   def test_fetch_resources_not_namespaced
     crd = mocked_cluster_resource_discovery
     kinds = crd.fetch_resources(namespaced: false).map { |r| r['kind'] }.uniq
-    assert_equal(kinds.length, 22)
+    assert_equal(kinds.length, 18)
```
Comment (jpfourny) [inline test/unit/cluster_resource_discovery_test.rb:17]: Less resources, since we removed some beta versions.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -40,14 +39,6 @@ jobs:
           kind_version: 'v0.11.1'
           kubernetes_version: '1.19.11'
           kind_image: 'kindest/node:v1.19.11@sha256:07db187ae84b4b7de440a73886f008cf903fcf5764ba8106a9fd5243d6f32729'
-        - ruby: '2.6.6'
```
Comment (jpfourny) [inline .github/workflows/ci.yml:43]: I don't we can support Ruby 2..6 after the Pysch 4 upgrade - there are incompatibilities with ActiveSupport.

File: test/integration-serial/serial_deploy_test.rb [hunk 13]
```diff
@@ -540,6 +540,9 @@ def test_resources_with_side_effect_inducing_webhooks_are_not_batched_server_sid
     result = deploy_global_fixtures("mutating_webhook_configurations", subset: %(ingress_hook.yaml))
     assert_deploy_success(result)
 
+    # Note: We have to mock `has_side_effects?`, since this won't be possible with K8s 1.22+.
```
[the remark this answers is not in the document]
Comment (JamesOwenHall) [reply test/integration-serial/serial_deploy_test.rb:543]: Will this test still be relevant once we drop support for k8s 1.21? If not, we should include a comment to remove the test once we get there.

[the remark this answers is not in the document]
Comment (JamesOwenHall) [reply test/integration-serial/serial_deploy_test.rb:543]: Same goes for the rest of the tests in this file and in `mutating_webhook_configuration_test.rb` that mock pre-1.22 behaviour.

**Review by JamesOwenHall — APPROVED** [review]

File: test/integration-serial/serial_deploy_test.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (jpfourny) [reply test/integration-serial/serial_deploy_test.rb:543]: Good idea - I will leave comments as a TODO to drop those tests (and the logic they are testing) after 1.21 is removed.

**Comment by jpfourny** [comment]
FYI: I am going to cut a new GEM version + README updates in a subsequent PR.

**Pull request merged**